MSC4205: Hashed moderation policy entities

Gnuxie opened the pull request:
[Rendered](https://github.com/Gnuxie/matrix-doc/blob/gnuxie/sha256-policy-entity/proposals/4205-sha256-policy-entity.md)

Implementations:

* Meowlnir: https://github.com/maunium/meowlnir/pull/4
* matrix-protection-suite: https://github.com/Gnuxie/matrix-protection-suite/pull/88
* Draupnir: https://github.com/the-draupnir-project/Draupnir/pull/761
    + Example protection: https://github.com/the-draupnir-project/Draupnir/pull/777

Signed-off-by: Gnuxie [Gnuxie@protonmail.com](mailto:Gnuxie@protonmail.com)

Gnuxie committed:
First draft.

Gnuxie committed:
Link up MSC

Gnuxie committed:
Fix typos from another reading.

Johennes commented on proposals/4205-sha256-policy-entity.md:
```diff
@@ -0,0 +1,94 @@
+
+## Proposal
+
+We therefore propose a new optional field `hashes` to the top level of
```
The existing `entity` field is currently required. Would you not also have to change it to optional to make the hash masquerading functional?

Gnuxie replied:
Yep that's right

Gnuxie replied:
Will this work for you? https://github.com/matrix-org/matrix-spec-proposals/commit/35049664369c6817f7794854a8cde2c5584d4f87#diff-62936d47ad39741850b1935807ef91072db5eccdbad8f3aba2f0bfd86eeeffb1R56-R58

Johennes replied:
I think this looks good, thank you. 👍

Johennes commented on proposals/4205-sha256-policy-entity.md:
```diff
@@ -0,0 +1,94 @@
+
+## Potential issues
+
+None noted.
```
I think hashing wouldn't help for glob-based rules?

Gnuxie replied:
Should I add anything else? https://github.com/matrix-org/matrix-spec-proposals/commit/35049664369c6817f7794854a8cde2c5584d4f87#diff-62936d47ad39741850b1935807ef91072db5eccdbad8f3aba2f0bfd86eeeffb1R63-R66

Johennes replied:
No, I think this looks great. 👍

Gnuxie committed:
Feedback from Johennes.

tulir commented on proposals/4205-sha256-policy-entity.md:
````diff
@@ -0,0 +1,102 @@
+```
+
+In this example, when a moderation tool encounters a new user, or a
+new policy, the tool will calculate the base64 encoded sha256
````
Many places in matrix use url-safe and/or unpadded base64, so if this is standard, it should probably be mentioned explicitly

Gnuxie commented on proposals/4205-sha256-policy-entity.md:
```diff
@@ -0,0 +1,102 @@
+
+Currently the content schema for `m.policy.rule.user` requires the
+`entity` field. In order for the `entity` field to be omitted when a
+hash has been provided, the entity field will have to become optional.
```
So, from my perspective there are policies in the wild that exist with omitted fields already, including the entity field. These were created by the now unmaintained spam police bot in an attempt to keep the context for a policy after unbanning a user.

The omission of entity, and the addition of the unstable hashes property, under the stable `m.policy.rule.room`, has not to my knowledge produced any significant problems for clients that have naive parsing of policy rules. Issues have also not yet been reported in implementation. But maybe there should be a requirement for a client to implement specific rendering for takedowns (gomuks might already) so that we can be sure.

Gnuxie replied:
> I think these events should have a distinct event type. Otherwise parsing the event is dependent on the recommendation, which both excludes ban recommendations from being hashed 

It's not tied to any recommendation. The schema of the policy is changed to make `entity` optional. That's the only change to parsing other than the addition of the `hashes` property. Parsing is not dependent on recommendation.

As for using the policy in an engine (not parsing). If `hashes` is present, then the hashes are used by the policy engine instead of the entity. If `hashes` is omitted then `entity` should be used instead and the policy should be treat as a literal.

Gnuxie committed:
Formatting.

Gnuxie committed:
Clarify encoding of the hash.

https://github.com/matrix-org/matrix-spec-proposals/pull/4205#discussion_r1992437074.
